### DISCUSSION TITLE
Code golf do codigo lindo

### DISCUSSION THREAD

**Discussion by munizart** [frontendbr/forum#550] [Discussão]
Tive essa ideia por conta da issue #543 do @fititnt.

A proposta aqui é o seguinte: Gerar o menor código _javascript_ que imprima de 1 a 5.

_Menor código_ vai ser contado pelo numero de caracteres.
_Imprimir_ pode ser qualquer saída do navegador/node.js

Vou tentar manter atualizado aqui o TOP3.
###TOP 3

1. @bernardodiasc: 5 chars
```javascript
12345
```
2. @GouveaHeitor: 6 chars
```
+12345
```
3. @munizart: 6 chars
```
0x3039
```

**Comment by munizart** [comment]
Minha tentativa:
```javascript
for(i=5;i--;)alert(5-i)
```
Com 23 caracteres

**Comment by paulinhoerry** [comment]
alert(1,2,3,4,5);

17 chars

**Comment by bernardodiasc** [comment]
`12345`

5 chars

**Comment by willycamargo** [comment]
Agora, falando de reaproveitamento de código e facilidade de manutenção. Se precisarmos alterar o código para imprimir de 1 a 50, no código do @munizart só é necessário alterar um valor e vai ter apenas um caracter a mais.

Enquanto nos outros dois códigos essa tarefa se torna maçante e improdutiva.

Claro que estamos falando de **escrever o menor código** e não de **escrever o código mais reaproveitável**, portanto a solução do @bernardodiasc será a menor haha.

**Comment by leandroruel** [comment]
quem concluir o desafio com sucesso receberá um contato telefônico do google. valendo!

**Comment by kazzkiq** [comment]
```js
[1,2,3,4,5].map(i=>alert(i))
```

~~28 caracteres. 😢~~

----
**EDIT:**
Consegui baixar pra 22 caracteres 😄 
```js
[1,2,3,4,5].map(alert)
```

**Comment by iamgbayer** [comment]
@paulinhoerry :joy::joy:

```js
alert(1,2,3,4,5)
```

16 chars

**Comment by danguilherme** [comment]
Pessoal, o `alert` só aceita 1 parâmetro, então vcs tão printando só `1`...

Seria um `alert(12345)`

**Comment by kivervinicius** [comment]
```
alert('1 a 5')
```
14 :dancer:

**Comment by Belline** [comment]
A resposta do @bernardodiasc está correta haha, isso gera um output no console, o que creio ser valido para o desafio, acho que é matematicamente impossivel ser menor do que isso hahaha

**Comment by htrgouvea** [comment]
Já que o @bernardodiasc se garantiu com 5 chars, aqui vou eu com os meus 7:

```
(12345)
```

**Comment by htrgouvea** [comment]
```
+12345
```

6 chars

**Comment by munizart** [comment]
`0x3039`
6 chars tambem

**Comment by danguilherme** [comment]
`030071`
6 chars também

**Comment by bernardodiasc** [comment]
1. `npm install -g express-generator`
2. `express imprimede1a5`
3. `cd imprimede1a5`
4. `npm install`
5. Crie novo arquivo `imprimede1a5/routes/numbers.js` contendo:

```js
var express = require('express');
var router = express.Router();

/* GET numbers listing. */
router.get('/', function(req, res, next) {
  res.json([{
    number: 1,
  }, {
    number: 2,
  }, {
    number: 3,
  }, {
    number: 4,
  }, {
    number: 5,
  }]);
});

module.exports = router;
```
6. Inclua `app.use('/numbers', numbers);` no arquivo `imprimede1a5/app.js`, logo abaixo de `app.use('/', index);`
7. Rode `PORT=3001 node bin/www` no terminal e deixe rodando
8. Abra outra janela do terminal e rode `npm install -g create-react-app`
9. depois `create-react-app client`
10. no `imprimede1a5/client/package.json` inclua `"proxy": "http://localhost:3001"`
11. rode `npm start` e deixe rodando, o browser vai abrir uma pagina pra voce com o app
12. no arquivo `imprimede1a5/client/src/App.js` altere o componente da seguinte forma:

```js
import React, { Component } from 'react';
import './App.css';

class App extends Component {
  state = {numbers: []}

  componentDidMount() {
    fetch('/numbers')
      .then(res => res.json())
      .then(numbers => this.setState({ numbers }));
  }

  render() {
    return (
      <div className="App">
        {this.state.numbers.map((each, index) =>
          <div key={index}>{each.number}</div>
        )}
      </div>
    );
  }
}

export default App;
```
13. salve o arquivo e confira o app no browser!

**Comment by victorperin** [comment]
```javascript
Array.from({length: 5}, (v, k) => k+1).join('');
```
48 chars 😝

**Comment by leandroruel** [comment]
@bernardodiasc ai você apelou euheuhe

**Comment by fititnt** [comment]
Se podemos usar transpilers, isso aqui deveria ser aceito.

5x3 = 15 bits (pode considerar mais, se usar separador de linhas)
5x7 = 35 bits é o tamanho da alternativa "12345"

![18195245_1474491142602362_1549566872_o](https://cloud.githubusercontent.com/assets/812299/25449181/5c330ec2-2a91-11e7-96a2-10ecf1d9e919.jpg)

**Comment by fititnt** [comment]
> @leandroruel @bernardodiasc ai você apelou euheuhe

Vocês ainda estão levando pela _zoeira_ a resposta dele, porém [ela é uma forma de _pensamento divergente_ e possivelmente é uma das que mais tem valor neste tópico para incitar (ou ensinar) outras formas de pensamento](https://github.com/frontendbr/forum/issues/543#issuecomment-297498650).

Por exemplo, o cara deu um guia passo a passo de como fazer algo usando ferramentes A+B+C, e pessoas se sentem inspiradas a tentar isso.

Por isso que eu estou dizendo: não deveriamos punir, mas incentivar respostas mais por criatividade do que "apenas eficiência", porque se travar demais as variáveis, você perde chance de ensinar e aprender coisas inovadoras.

**Comment by leandroruel** [comment]
ah cara, eu já entendi o seu ponto, mas aqui é BR.

**Comment by bernardodiasc** [comment]
![](https://pbs.twimg.com/profile_images/567465842120863744/-VAkNxNp.jpeg)

**Comment by podrivo** [comment]
Esse tópico me lembrou essa palestrinha: https://www.destroyallsoftware.com/talks/wat

**Comment by fititnt** [comment]
>> @fititnt Vocês ainda estão levando pela _zoeira_ a resposta dele, porém [ela é uma forma de _pensamento divergente_ e possivelmente é uma das que mais tem valor neste tópico para incitar (ou ensinar) outras formas de pensamento](https://github.com/frontendbr/forum/issues/543#issuecomment-297498650).
 
>  @leandroruel ah cara, eu já entendi o seu ponto, mas aqui é BR.

> @bernardodiasc  
> ![huehuehuehue](https://camo.githubusercontent.com/d31412bff0f9bfb31eb48697460476a08c517fb2/68747470733a2f2f7062732e7477696d672e636f6d2f70726f66696c655f696d616765732f3536373436353834323132303836333734342f2d56416b4e784e702e6a706567)

Humm...

1. Parece que "zuera" e "huehuehue", empiricamente falando, é uma metáfora para "pensamento divergente". Ou seja, o objetivo final -- aprender a a aprender -- acaba sendo o mesmo.
2. "Pensamento divergente" é uma das bases para criatividade (que implicitamente pode ser melhorada tornando indivíduo mais sobre os próprios mecanismo de aprendizagem -- metacognição). [E criatividade é essencial para empreendedorismo, e até sucesso profissional, tão ou mais importante que coeficiente de inteligência](http://www.businessnewsdaily.com/5813-creativity-in-entrepreneurship.html).
3. [O Brasil está perdendo muito dinheiro em inovação tecnológica, e até mesmo gerando infelicidade, de trabalhadores; escolas não estão preparando pro mercado de trabalho](https://github.com/frontendbr/forum/issues/543#issuecomment-297227710); é comum pessoas falarem que "universidade está me deixando pior do que não fazê-la", sendo recorrente indiretamente afirmações que levam a induzir ainda menos criatividade nos trabalhadores, em especial aqueles que querem seguir muito regras pre-definidas

**Parece ter lógica permitir, ou mesmo induzir, uso de [meta-piadas/"zuera"/"huehuehue"](https://en.wikipedia.org/wiki/Meta-joke) é sim, uma das formas extremamente eficiêntes e otimizadas para aprender a gerar riqueza para si e para o Brasil, mesmo que o individuo não tenha ideia disso** 🤔

**Comment by vitocaio** [comment]
1 a 5

3 chars

**Comment by diessica** [comment]
NOPE!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
![image](https://cloud.githubusercontent.com/assets/5303585/25503698/a80e68b0-2b71-11e7-860e-bef876a6c05c.png)

**Comment by fititnt** [comment]
E se o @vitocaio não errou, mas considerou no calculo uma metafora para LLOC a um nível extremo de texto?

Além disso, ele _pode_ ter considerado na contagem dele caracteres imprimíveis. Maioria dos editores de texto e de código que tem alguma função de contar caracteres costumam ignorar espaços em branco.

O simples fato de, em vez de dizer "sua resposta esta errada" parar pra pensar se faz sentido mesmo quando ouvir algo absurdo pode levar a aprender ou deduzir coisas bem profundas.

PS.: Por essa lógica, o "**javascript que imprima de 1 a 5**" viraria "**de 1 a 5**" que voltaria a ser 5 caracteres imprimíveis. 

----

[![captura de tela de 2017-04-27 18-44-38](https://cloud.githubusercontent.com/assets/812299/25505747/9b00932a-2b79-11e7-9cc2-4324e79b0226.png)
http://www.projectcodemeter.com/cost_estimation/help/GL_lloc.htm
](http://www.projectcodemeter.com/cost_estimation/help/GL_lloc.htm)

**Comment by danguilherme** [comment]
![facepalm](https://cloud.githubusercontent.com/assets/1488378/25508492/d0a6fa28-2b88-11e7-879e-5ec203c97d37.jpg)

Queria ter essa paciência que vc tem com trolls...

**Comment by fititnt** [comment]
> @danguilherme Queria ter essa paciência que vc tem com trolls...

Tenho paciência com eles assim como tenho com você. Em três prints te mostro que a "melhor resposta" parecia trollagem, e, para esse excercício proposto pelo @munizart, ela não é.

As duas primeiras são da postagem #543 e a ultima é uma bem aceita aqui.

-----

![captura de tela de 2017-04-27 22-43-06](https://cloud.githubusercontent.com/assets/812299/25511111/37ca4d7e-2b9b-11e7-9149-0eaa70a01d4c.png)

-----

![captura de tela de 2017-04-27 22-44-13](https://cloud.githubusercontent.com/assets/812299/25511114/3ff437e4-2b9b-11e7-9513-fc16679780fc.png)

-----

![captura de tela de 2017-04-27 22-45-04](https://cloud.githubusercontent.com/assets/812299/25511119/4868a798-2b9b-11e7-9fd4-bac89f7e2183.png)


-----

Ensinar a pensar diferente é diferencial inovativo. É a diferença entre quem segue tendência, e quem as cria.

![d08aa9755aa59d9e164c30fd797d27bd](https://cloud.githubusercontent.com/assets/812299/25511493/76028c6c-2b9d-11e7-9cfa-4781d8755b3c.jpg)

![steve-jobs-quotes-on-leadership-3](https://cloud.githubusercontent.com/assets/812299/25511503/8745f09a-2b9d-11e7-9503-a21fa3e1b79c.jpg)

**Comment by bernardodiasc** [comment]
🎉 @wmsbill 🎉
🎉💃 :man_dancing: 🎉
🎉 parabéns, voce ganhou 1 iPad! 🎉  
👏😄👏😄👏😄👏😄👏😄👏😄👏😄

**Comment by wmsbill** [comment]
Queria agradecer a todos que me apoiaram até esse ponto da minha vida.

Obrigado a todos!

**Comment by fititnt** [comment]
TL;DR: desculpa

----

Pessoal, eu explicitamente **vou pedir desculpas** pelo comentário https://github.com/frontendbr/forum/issues/550#issuecomment-297887328 pelo seguinte motivo: **eu fiz uma comparação entre pessoas A e B, implicitamente com referência ao nome, onde o comportamento de uma seria melhor do que o de outra e que isso poderia afetar futuro, como resposta a pessoa C que me deixou levemente estressado**

O que possivelmente me levou a fazer isso 9 dias atras:

- Eu naquele momento acreditei que estava elogiando a pessoa B, e como resposta a pessoa C que fez um comentário leve, não especialmente negativo, fui incisivo nas minhas afirmações (importante: eu sabia que estava sendo incisivo)
- Eu acreditava que deveria fazer isso; não estava especialmente irritado, alias, eu nem estava irritado, o que pode ter me feito não ler mais de uma vez e repensar no que iria postar (importante: é comum em situações que parecem mais tensas a gente dosar mais argumentos)
- Eu deixei de considerar o impacto negativo na pessoa A por ter interagido positivamente com ele em outros momentos (importante: é comum entre homens fazer _piadas_ com colegas sem saber que estão se ofendendo)

O que possivelmente aconteceu, que **não era esperado**, isto é, efeito colateral indesejado

- Eu provavelmente desestimulei A, que pode ter usado ironia como forma de defesa
- Possivelmente outras pessoas podem copiar meu comportamento

## Algumas considerações heurísticas que eu percebi que pode ser útil para não machucar pessoas que não quer machucar

> Esse vídeo, que possui legendas em português, fala sobre como a mente pode pregar peças (tanto pelo lado positivo quanto negativo) https://www.youtube.com/watch?v=R-sVnmmw6WY&index=15&list=PL8dPuuaLjXtOPRKzVLY0jJY-uHOH9KVU6

1. Toda conversa por texto tem problemas de comunicação, e pode ser compreendida de forma bem diferente; não ver expressão facial, corpo, etc afeta julgamento (aka _nunca se discute relacionamento por texto, melhor webcam, sério_)
2. Sempre que for possível elogiar uma pessoa ou um comportamento sem criticar outra, mesmo se for pessoa conhecida, ou que pareça não se importar, parece mais seguro não ofender
3. Mesmo quando parecer que alguém está nos atacando ([vide perceber é acreditar](https://www.youtube.com/watch?v=n46umYA_4dM&t=23s&list=PL8dPuuaLjXtOPRKzVLY0jJY-uHOH9KVU6&index=7)), mesmo que sejam várias, parece mais seguro não necessariamente ignorar, mas considerar que um ataque não valida o outro
4. O uso de ironias (tanto por nós, como por outros), pode dar indícios de pensamentos não irônicos, e isso deve ser considerado em momento futuro para se reavaliar.

Não quer dizer que o que eu disse nessa postagem aqui é verdade; o ponto é que quebrar o processo de pensamento é útil pra se entender e entender os outros e melhorar [teoria da mente](https://pt.wikipedia.org/wiki/Teoria_da_mente) ao ponto de deduzir coisas mesmo quando pessoas não falam aka efeito _"nooossa-você-me-entende"_.
